src/db_adapters/access_keys.rs

```diff
@@ -15,107 +14,71 @@ use crate::{metrics, models};
-                                created_by_receipt_id: None,
-                                deleted_by_receipt_id: Some(receipt.receipt_id.to_string()),
-                                // this is a workaround to avoid additional struct with optional field
-                                // permission_kind is not supposed to change on delete action
-                                permission_kind: models::enums::AccessKeyPermission::FullAccess,
-                                last_update_block_height: block_height.into(),
-                            });
-                    }
-                    near_indexer::near_primitives::views::ActionView::Transfer { .. } => {
-                        if receipt.receiver_id.len() != 64usize {
-                            continue;
-                        }
-                        if let Ok(public_key_bytes) = hex::decode(receipt.receiver_id.as_ref()) {
-                            if let Ok(public_key) =
-                                near_crypto::ED25519PublicKey::try_from(&public_key_bytes[..])
-                            {
-                                access_keys.insert(
-                                    (near_crypto::PublicKey::from(public_key.clone()).to_string(), receipt.receiver_id.to_string()),
-                                    models::access_keys::AccessKey::from_action_view(
-                                        &near_crypto::PublicKey::from(public_key.clone()),
-                                        &receipt.receiver_id,
-                                        &near_primitives::views::AccessKeyView {
-                                            nonce: 0,
-                                            permission: near_primitives::views::AccessKeyPermissionView::FullAccess
-                                        },
-                                        &receipt.receipt_id,
-                                        block_height,
-                                    ),
-                                );
-                            }
-                        }
-                    }
-                    _ => continue,
+    for state_change in state_changes {
+        if let near_primitives::views::StateChangeCauseView::ReceiptProcessing { receipt_hash } =
+            state_change.cause
+        {
+            match &state_change.value {
+                near_primitives::views::StateChangeValueView::AccountDeletion { account_id } => {
+                    deleted_accounts.insert(account_id.to_string(), receipt_hash.to_string());
+                    access_keys
+                        .iter_mut()
+                        .filter(|((_, receiver_id), _)| receiver_id == &account_id.to_string())
+                        .for_each(|(_, access_key)| {
+                            access_key.deleted_by_receipt_id = Some(receipt_hash.to_string());
+                        });
                 }
+                near_primitives::views::StateChangeValueView::AccessKeyUpdate {
```

Comment:
This is triggered when an access key is created but the `Update` makes me wonder what else triggers it, is `insert`ing here going to be safe?

Reply:
Oh, good catch! The keys are updated with every transaction they are used to sign transactions, since nonce value gets updated. This means that we need to properly distinguish creation from these updates. Oops, we need to take some time to think through it. It seems there will be a need to make a hybrid approach of the previous one and the new one to catch access keys creation properly

Reply:
The solution is that once we observe the key update state change, check the receipt actions for the matching AddKey or Transfer (in case if the receiver account id is implicit, and if the updated key matches the implicit access key, `decode_hex(receipt.receiver_account_id) == updated_access_key.public_key.as_bytes()`).

Reply:
Disregard my previous comments. It will be TransactionProcessing event that updated the nonce, so we are good here with only handling ReceiptProcessing state changes. No changes needs to be made here

Reply:
Awesome 👍🏽